docs/HighLevelDesignForGraphProgramExtraction.md

```diff
@@ -0,0 +1,70 @@
+
+### Single TF Graph
+
+We construct the TF graph by _walking through_ the Swift program and creating a program with the exact same structure, but keeping only the TF-relevant parts. This means that we start at the top-level Swift code, that will be executed when we execute the Swift program. For now, we ignore functions as that will be discussed when we tackle point 3.
+
+For all top-level code we check where `Tensor`s are used and how and we built a TF graph, based on the methods outlined in the GPE whitepaper. `Tensor`s that are created on the Swift-side are handled by the TF graph in the way described in the section tackling point 2. After building the TF graph in this manner, we collect all sink nodes in the TF graph, whether those are nodes that send tensors back to the Swift program, or simply nodes that only have a side-effect on the TF side, and we create a new TF `group` node that we call the `sink` node. This means that once the `sink` node finishes executing in TF, all TF-related computation for the current Swift for TF program will have finished executing and so it can be thought of as the TF _endpoint_.
+
+We have ignored functions that use TF for now, but we will discuss those in the section devoted to point 3.
+
+### Single TF Session
+
+Given the single TF graph we have built for the current Swift for TF program, we can then go ahead and add the following in the beggining of the top-level Swift program code (on the Swift side):
+
+1. Create a new TF session (with any desirable configuration).
+2. Add a call to `session.run(sink)`.
+
+This last call will start execution of the TF program, so that it is running in parallel with the Swift program. It allows all parts of the graph that do not depend on communicating with Swift at any point in time, to be executed in parallel with the main Swift program. The fact that we end up with a single TF session means that it becomes easy to manage the session configuration and to profile the TF program performance. Furthermore, no two TF sessions have to fight for the same resources. It also makes this program slicing approach cleaner in that there is only two programs running at any point in time, a single Swift program and its corresponding TF program.
+
+## Tackling Point 2: Communication Between Swift and TF
+
+Communication between Swift and TF can be handled in a clean manner, by following the same principles.
+
+1. ___Swift -> TF:___ The TF program deals with `Tensor` objects. Therefore, whenever a `Tensor` object is constructed on the Swift side, and later used by TF ops that require tensors, a corresponding `SendToTF` op is added to the graph. This actually consists of two ops: (i) an `Enqueue` op, and a (ii) `Dequeue` op to a size 1 queue. The `Dequeue` op will be used as the `Tensor` object representation for TF ops consuming it later on, and a `session.run(...)` call will be added to the Swift side, enqueueing the constructed tensor that will be sent to TF.
```

Comment:
note "size 1 queue" can be relaxed and tuned. a larger sized queue could help with throughput, but might hurt latency (say we and sending / receiving a tensor value within a loop, we can send the values across many iterations without them being consumed, in some cases -- it's a producer/consumer buffer)

Reply:
That's a good point! I will remove the size 1 constraint. :)